packages/libs/error-stack/src/iter.rs

```diff
@@ -47,14 +101,14 @@ impl fmt::Debug for Frames<'_> {
 /// Use [`Report::frames_mut()`] to create this iterator.
 #[must_use]
 pub struct FramesMut<'r> {
-    current: Option<NonNull<Frame>>,
+    current: Vec<NonNull<Frame>>,
     _marker: PhantomData<&'r mut Frame>,
 }
 
 impl<'r> FramesMut<'r> {
     pub(crate) fn new<C>(report: &'r mut Report<C>) -> Self {
         Self {
-            current: Some(NonNull::from(report.frame_mut())),
+            current: report.frames.iter_mut().map(NonNull::from).rev().collect(),
```

Comment:
We might want to store the iterator rather than a collection. This makes the traversal lazy rather then precalculating everything.
You could also apply this to `Frames` (remove the const, we have a breaking change anyway and there is no real need for a `const new`)

Reply:
Yep makes sense, frames initially was the same as FramesMut but due to const limitations I needed to get creative, if it's save to remove const I'll gladly rework

Reply:
I have been able to convert the `Frames` to an iterator based approach, but I am currently failing with `NonNull` and FramesMut, due to the fact that I cannot properly type the attribute, without using `dyn` (it's late in my timezone, so maybe sleeping over the whole thing might help^^).

Reply:
@TimDiekmann I have been able to get a version working for `FramesMut`, but I am unsure about the soundness (I am pretty new to unsafe rust). Could you review it and make sure I didn't do anything wrong?

Reply:
Sure, I'll take a look, but probably not before tomorrow or Monday.

Reply:
Would this PR would your life make a little bit easier? 🙂 

- https://github.com/hashintel/hash/pull/774

Reply:
https://github.com/hashintel/hash/blob/45db0a71259d25adae5b100f460e08bda8ead133/packages/libs/error-stack/src/iter.rs#L138-L142

I'm actually not 100% sure if this is sound. My guess would be yes. The `IterMut` will hold a mutable reference to all remaining frames and you `pop` the reference from the vector in `fn next()`. I'm not sure, because the `next()` function isn't really easy to read so it's hard to reason about. In any case, the SAFETY section on the `unsafe` block is not reflecting the actual unsafe code.

A suggestion which should make this easier:
- The iterator holds frames to visit.
    ```rust
    pub struct FramesMut<'r> {
        to_visit: Vec<NonNull<Frame>>,
        _marker: PhantomData<&'r mut Frame>,
    }
    ```
- It's populated with all root frames (should probably take `&[Frame]` instead of `&[Report]`). It's pushed in reverse order so it's easier to pop a `Frame` and add new to the back:
    ```rust
    pub(crate) fn new<C>(report: &'r mut Report<C>) -> Self {
        Self {
            to_visit: report.frames.iter_mut().rev().map(NonNull::from).collect(),
            _marker: PhantomData,
        }
    }
    ````
- The `next` implementation would look like this:
    ```rust
    unsafe {
        let current = self.to_visit.last_mut()?.as_mut();
        self.to_visit
            .extend(current.sources_mut().iter_mut().rev().map(NonNull::from));
        Some(current)
    }
    ```

Do you have a test case for this? The tree shown in the documentation should probably also used for testing and then let `miri` decide, if this is sound 🙂

Reply:
Yep, like the Frames implementation i tried to make it as lazy as possible, which then resulted in the usage of raw pointer. You are correct the comment doesn't reflect the actual code anymore. There are two test cases at the bottom (both Frames and FramesMut) which take the tree in the doc comment and verify that it works.

Reply:
I tested the whole implementation against `miri`, and it seems like it is indeed sound. I have also rewritten the `SAFETY` comment.

If we want to have an eagerly evaluated list instead, I can roll back to the previous version, which made use of `Vec` and `NonNull`, instead of a list of iterators.

Reply:
I converted this to an Iterator, instead of the `Vec<NonNull<Frame>>`, after your comment two days ago

> We might want to store the iterator rather than a collection. This makes the traversal lazy rather then precalculating everything.

The code snippet you showed me was actually pretty much the implementation before the change^^ so I can roll back if that makes it clearer.

Reply:
Almost, I asked you to change the inner `Vec` to an iterator 🙂
Given, that I was *almost* sure it's sound and `miri` is running fine, I'm satisfied, so you may want to leave it as it is. My solution would have been an alternative, but I don't claim that it would have been better.

Your latest commit made the `fn next()` more readable, and I like the fact it can be used for both iterators. My approach probably would have needed `unsafe` in the shared iterator as well, which is bad. The SAFETY section looks good. Only two small nits: Can you remove the `_marker` as the lifetime is now captured in the iterator? And `(&mut *frame)` can be changed to `(*frame)`.

Reply:
I think if we do micro-optimization, this would probably be a good spot to look at again, but without any performance baseline, it's pointless IMO. I don't see an issue in performance here currently as this is most likely the error **handling** part, not the error **creation** part.